File: Plugins/AudioPlayer/src/ios_AudioPlayer/AudioPlayer/AWSS3Cache.swift
//
//  AWSS3Cache.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/9/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//
// The read function will read cache and return it if present and unexpired, then it will access online
// And return the result, saving it in cache after it is returned.
//
//

import Foundation
import AWS

class AWSS3Cache {
    
    let cacheDir: URL
    
    init() {
        let homeDir: URL = URL(fileURLWithPath: NSHomeDirectory(), isDirectory: true)
        let libDir: URL = homeDir.appendingPathComponent("Library")
        self.cacheDir = libDir.appendingPathComponent("Caches")
    }
    
    deinit {
        print("***** Deinit AWSS3Cache *****")
    }
    
    func read(s3Bucket: String, s3Key: String, expireInterval: TimeInterval,
              getComplete: @escaping (_ data: Data?) -> Void) {
        let localKey = self.getLocalKey(s3Bucket: s3Bucket, s3Key: s3Key)
        let path: URL = self.cacheDir.appendingPathComponent(localKey)
        let data: Data? = self.readCache(path: path, expireInterval: expireInterval)
        if data != nil {
            getComplete(data)
        } else {
            self.readAWSS3(s3Bucket: s3Bucket, s3Key: s3Key, filePath: path, getComplete: getComplete)
        }
    }
    
    private func readCache(path: URL, expireInterval: TimeInterval) -> Data? {
        print("Path to read \(path)")
        do {
            let data = try Data(contentsOf: path, options: [])
            if (self.isFileExpired(filePath: path, expireInterval: expireInterval)) {
                print("File has expired in AWSS3Cache.readCache")
                return nil
            } else {
                return data
            }
        } catch( let err ) {
            print("Error occur in AWSS3Cache.readCache \(err)")
            return nil
        }
    }
    
    private func readAWSS3(s3Bucket: String, s3Key: String, filePath: URL,
                getComplete: @escaping (_ data: Data?) -> Void) {
        AwsS3.shared.downloadData(
            s3Bucket: s3Bucket,
            s3Key: s3Key,
            complete: { error, data in
                if let err = error {
                    print("Error accessing S3 in MetaDataCache \(err)")
                    getComplete(nil)
                }
                else {
                    getComplete(data)
                    do {
                        try data?.write(to: filePath)
                    } catch( let writeErr ) {
                        print("Error while writing file in AWSS3Cache.readAWSS3 \(writeErr)")
                    }
                }
            }
        )
    }

    private func getLocalKey(s3Bucket: String, s3Key: String) -> String {
        return(s3Bucket + "_" + s3Key)
    }
    
    private func isFileExpired(filePath: URL, expireInterval: TimeInterval) -> Bool {
        do {
            let dictionary = try FileManager.default.attributesOfItem(atPath: filePath.path)
            let creationDate = dictionary[FileAttributeKey.modificationDate] as? Date
            if let creation = creationDate {
                print("creationDate \(creation)")
                let interval = abs(creation.timeIntervalSinceNow)
                return interval > expireInterval
            } else {
                return true
            }
        } catch let err {
            print("Error getting creation date \(err) ON FILE \(filePath)")
            return true
        }
    }
}

File: Plugins/AudioPlayer/src/android_AudioPlayer/app/src/main/java/com/shortsands/audioplayer/AWSS3Cache.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.content.Context;
import android.util.Log;
import com.shortsands.aws.AwsS3;
import com.shortsands.aws.DownloadFileListener;
import com.shortsands.io.FileManager;
import java.io.File;
import java.util.Date;

class AWSS3Cache extends DownloadFileListener {

    private static final String TAG = "AWSS3Cache";

    private final Context context;
    private final CompletionHandler completionHandler;
    private final File cacheDir;

    AWSS3Cache(Context context, CompletionHandler handler) {
        super();
        this.context = context;
        this.completionHandler = handler;
        this.cacheDir = context.getCacheDir();
    }

    //deinit {
    //    print("***** Deinit AWSS3Cache *****")
    //}

    void read(String s3Bucket, String s3Key, int expireInterval) {
        String localKey = this.getLocalKey(s3Bucket, s3Key);
        File filePath = new File(this.cacheDir, localKey);
        String data = this.readCache(filePath, expireInterval);
        if (data != null) {
            this.completionHandler.completed(data, this);
        } else {
            AwsS3.shared().downloadFile(s3Bucket, s3Key, filePath, this);
            //this.downloadFile(s3Bucket, s3Key, filePath, this);
            Log.d(TAG, "**** AWSS3Cache performed downloadFile");
        }
    }

    private String readCache(File path, int expireInterval) {
        Log.d(TAG, "Path to read " + path.toString());
        String data = FileManager.readTextFully(path); // does not throw
        if (data != null) {
            if (this.isFileExpired(path, expireInterval)) {
                Log.d(TAG, "File has expired in AWSS3Cache.readCache");
                return null;
            } else {
                return data;
            }
        } else {
            return null;
        }
    }

    private String getLocalKey(String s3Bucket, String s3Key) {
        return(s3Bucket + "_" + s3Key);
    }

    private boolean isFileExpired(File filePath, int expireInterval) {
        long modificationDate = filePath.lastModified();
        Log.d(TAG, "modification date " + modificationDate);
        long now = new Date().getTime();
        Log.d(TAG, "now date " + now);
        long interval = Math.abs(now - modificationDate);
        Log.d(TAG, "interval " + interval);
        return (interval > expireInterval);
    }

    /**
     * The onComplete method and onError method are part of the DownloadFileListener interface.
     * One of them is called by AwsS3.downloadFile when the data has been downloaded.
     * @param id
     */
    @Override
    protected void onComplete(int id) {
        Log.d(TAG, "**** Inside AWSS3Cache.onComplete");
        super.onComplete(id);
        //this.results = this.file; // This is done by super
        String data = FileManager.readTextFully(this.results);
        //String data = this.results;
        this.completionHandler.completed(data, this);
    }

    @Override
    public void onError(int id, Exception e) {
        Log.d(TAG, "**** Inside AWSS3Cache.onError");
        super.onError(id, e);
        //Log.e(TAG, "Error: " + e.toString() + " on " + this.file.getAbsolutePath()); // This is done by super
        this.completionHandler.failed(e, this);
    }
}
